api/task_create_test.go

```diff
@@ -0,0 +1,293 @@
+			var actual oapigen.ErrorResponse
+			err = decoder.Decode(&actual)
+			require.NoError(t, err)
+
+			expected := generateErrorResponse("", tc.message)
+			assert.Equal(t, expected, actual)
+		})
+	}
+}
+
+func TestTaskLifeCycleHandler_CreateTask_InternalError(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		name       string
+		taskName   string
+		request    string
+		statusCode int
+		message    string
+		response   taskResponse
+	}{
+		{
+			name:       "already exists",
+			taskName:   testTaskName,
+			request:    testCreateTaskRequest,
+			message:    "error initializing new task, mock error",
+			statusCode: http.StatusInternalServerError,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			c, d := generateTaskLifecycleHandlerTestDependencies()
+
+			d.On("InitTask", mock.Anything).Return(errors.New("this is an error"))
+			conf := driver.TaskConfig{
+				Name: tc.taskName,
+			}
+			task, err := driver.NewTask(conf)
+			require.NoError(t, err)
+			d.On("Task").Return(task)
+
+			resp := runTestCreateTask(t, c, tc.statusCode, tc.request)
+
+			// Task should not be added to the list
+			_, ok := c.drivers.Get(tc.taskName)
+			require.False(t, ok)
+
+			// Only one event should be registered
+			checkTestEventCount(t, tc.taskName, c.store, 1)
+
+			// Check response
+			decoder := json.NewDecoder(resp.Body)
+			var actual oapigen.ErrorResponse
+			err = decoder.Decode(&actual)
+			require.NoError(t, err)
+
+			expected := generateErrorResponse("", "error initializing new task, this is an error")
+			assert.Equal(t, expected, actual)
+		})
+	}
+}
+
+func generateExpectedResponse(t *testing.T, req string) taskResponse {
+	var treq taskRequest
+	err := json.Unmarshal([]byte(req), &treq)
+	require.NoError(t, err)
+
+	tc, err := treq.ToConfigTaskConfig(config.DefaultBufferPeriodConfig(), "sync-task")
+	require.NoError(t, err)
+
+	tresp := taskResponseFromConfigTaskConfig(tc, "")
+	require.NoError(t, err)
+
+	return tresp
+}
+
+func generateErrorResponse(requestID, message string) oapigen.ErrorResponse {
+	errResp := oapigen.ErrorResponse{
+		Error: oapigen.Error{
+			Message: message,
+		},
+		RequestId: oapigen.RequestID(requestID),
+	}
+
+	return errResp
+}
+
+func generateTaskLifecycleHandlerTestDependencies() (TaskLifeCycleHandlerConfig, *mocks.Driver) {
+	// Create dependencies
+	// 1) Event Store
+	store := event.NewStore()
+
+	// 2) drivers list
+	drivers := driver.NewDrivers()
+
+	// 3) bufferPeriod
+	bufferPeriod := config.DefaultBufferPeriodConfig()
+
+	// 4) Working Directory
+	workingDir := "sync-task"
+
+	// 5) createNewTaskDriver func
+	d := new(mocks.Driver)
+	createNewTaskDriver := func(taskConfig config.TaskConfig) (driver.Driver, error) {
+		return d, nil
+	}
+
+	c := TaskLifeCycleHandlerConfig{
+		store:               store,
+		drivers:             drivers,
+		bufferPeriod:        bufferPeriod,
+		workingDir:          workingDir,
+		createNewTaskDriver: createNewTaskDriver,
+	}
```

Comment:
optional suggestion - the dependencies created above are pretty straightforward such that the comments don't add much additional information. would you consider something like below? i find that's still fairly readable and reduces line count


```
	d := new(mocks.Driver) // this is the same
  	createNewTaskDriver := func(taskConfig config.TaskConfig) (driver.Driver, error) {
  		return d, nil
  	}
	c := TaskLifeCycleHandlerConfig{
  		store:               event.NewStore(),
  		drivers:             driver.NewDrivers(),
  		bufferPeriod:        config.DefaultBufferPeriodConfig(),
  		workingDir:          "sync-task",
  		createNewTaskDriver: createNewTaskDriver,
  	}
```

Reply:
yep, that's fine. This was more a checklist for me when I was creating this, but you're right, now that it's done it makes sense to refactor it.